src/soot/jimple/toolkits/callgraph/OnFlyCallGraphBuilder.java

```diff
@@ -414,57 +527,375 @@ public void processReachables() {
+						Iterator<SootMethod> mIt = getPublicMethodIterator(baseClass, reachingTypes, methodSizes, mustNotBeNull);
+						while(mIt.hasNext()) {
+							SootMethod sm = mIt.next();
+							cm.addVirtualEdge(ics.container(), ics.stmt(), sm, Kind.REFL_INVOKE, null);
+						}
+					}
+				}
+			}
+		}
+	}
+    
+    private void resolveStaticTypes(Set<Type> s, InvokeCallSite ics) {
+    	ArrayTypes at = ics.reachingTypes();
+		for(Type bType : s) {
+			SootClass baseClass = ((RefType)bType).getSootClass();
+			Iterator<SootMethod> mIt = getPublicMethodIterator(baseClass, at);
+			while(mIt.hasNext()) {
+				SootMethod sm = mIt.next();
+				cm.addVirtualEdge(ics.container(), ics.stmt(), sm, Kind.REFL_INVOKE, null);
+			}
+		}
+	}
+	private Iterator<SootMethod> getPublicMethodIterator(SootClass baseClass, final ArrayTypes at) {
+		return new AbstractMethodIterator(baseClass) {
+			@Override
+			protected boolean acceptMethod(SootMethod m) {
+				if(!at.possibleSizes.contains(m.getParameterCount())) {
+					return false;
+				}
+				for(int i = 0; i < m.getParameterCount(); i++) {
+					if(at.possibleTypes[i].isEmpty()) {
+						continue;
+					}
+					if(!isReflectionCompatible(m.getParameterType(i), at.possibleTypes[i])) {
+						return false;
+					}
+				}
+				return true;
+			}
+			
+		};
+	}
+	
+	private PrimType[] narrowings(PrimType f) {
+		if(f instanceof IntType) {
+			return INT_NARROWINGS;
+		} else if(f instanceof ShortType) {
+			return SHORT_NARROWINGS;
+		} else if(f instanceof LongType) {
+			return LONG_NARROWINGS;
+		} else if(f instanceof ByteType) {
+			return BYTE_NARROWINGS;
+		} else if(f instanceof FloatType) {
+			return FLOAT_NARROWINGS;
+		} else if(f instanceof BooleanType) {
+			return BOOLEAN_NARROWINGS;
+		} else if(f instanceof DoubleType) {
+			return DOUBLE_NARROWINGS;
+		} else if(f instanceof CharType) {
+			return CHAR_NARROWINGS;
+		} else {
+			throw new RuntimeException("Unexpected primitive type: " + f);
+		}
+	}
+
+	private boolean isReflectionCompatible(Type paramType, Set<Type> reachingTypes) {
+		/* 
+		 * attempting to pass in a null will match any type (although attempting to pass it
+		 * to a primitive arg will give an NPE)
+		 */
+		if(reachingTypes.contains(NullType.v())) {
+			return true;
+		}
+		if(paramType instanceof RefLikeType) {
+			for(Type rType : reachingTypes) {
+				if(fh.canStoreType(paramType, rType)) {
+					return true;
+				}
+			}
+			return false;
+		} else if(paramType instanceof PrimType) {
+			PrimType primType = (PrimType) paramType;
+			/*
+			 * It appears, java reflection allows for unboxing followed by widening, so if 
+			 * there is a wrapper type that whose corresponding primitive type
+			 * can be widened into the expected primitive type, we're set
+			 */
+			for(PrimType narrowings : narrowings(primType)) {
+				if(reachingTypes.contains(narrowings.boxedType())) {
+					return true;
+				}
+			}
+			return false;
+		} else {
+			// impossible?
+			return false;
```

Comment:
I could be missing something here but I think this branch should not be reachable.